### PR TITLE
Feat: More ai tools

### PR DESCRIPTION
**Pull request by Scoppio**
# What this PR do?

It adds the the concept of cost function and a parameter optimization that uses gradient descent with some spicyness to it.

This optimizer can receive multiple game action logs and turn them all into a single training dataset which it will them try to optimize for.

Right now it accepts up to 3 different cost functions as arguments to optimze for, the first is our current princess, the other two are experiments based on future implementations (and the current experimental Utility path ranker).

It implements a command line access to it and also a gradle task which requires arguments to run.

# It's amazing!?

Not so much, this is just a tool to help us develop the parameters for more complex AIs, however, now we depend heavily in the cost function that is still not up to the task, or simply isn't implemented yet.

The game actions log too isn't a perfect tool and because of that may limit us a little bit in what we can achieve with the sinthetic cost functions right now. (But we will make do)

### CODE REVIEW

File: megamek/src/megamek/utilities/ai/BasicPathRankerCostFunction.java [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+        var crowdingEnemies = nonAlliedUnits(unitAction, currentUnitStates).stream()
+            .map(UnitState::position)
+            .filter(c -> c.distance(position) <= closingDistance)
+            .count();
+        double friendsCrowdingTolerance = antiCrowdingFactor * crowdingFriends;
+        double enemiesCrowdingTolerance = antiCrowdingFactor * crowdingEnemies;
+        return friendsCrowdingTolerance + enemiesCrowdingTolerance;
+    }
+
+    private static double calculateMovementMod(UnitAction unitAction, Map<Integer, UnitState> currentUnitStates, BehaviorSettings behaviorSetting) {
```
[the remark this answers is not in the document]
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/BasicPathRankerCostFunction.java:205]: Solved

File: megamek/src/megamek/common/Entity.java [hunk 2]
```diff
@@ -2574,7 +2574,7 @@ public boolean isOnAtmosphericGroundMap() {
      * @return True if this is an airborne aircraft on a ground map.
      */
     public boolean isAirborneAeroOnGroundMap() {
-        return isAero() && isAirborne() && getGame().getBoard().onGround();
+        return isAero() && isAirborne() && getGame() != null && getGame().getBoard().onGround();
```
Comment (Sleet01) [inline megamek/src/megamek/common/Entity.java:2577]: Rather than adding a bunch of null checks, do we want to think about creating a new lightweight Game class just for AI training?

File: megamek/src/megamek/utilities/PrincessFineTuning.java [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+        for (int i = 2; i < args.length; i++) {
+            File file = new File(args[i]);
+            if (!file.exists()) {
+                System.out.println("File not found: " + file);
+                return;
+            }
+            datasetParser.parse(file);
+            counter++;
+            System.out.println("Loaded dataset " + counter + " of " + (args.length - 2));
+        }
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/PrincessFineTuning.java:58]: I don't think we want to mix arg parsing and main function logic.  I would highly recommend:
1. Bringing in a prebuilt argument parsing library like `argparse4j`,
2. Break arg parsing (and argument values) into a separate class, maybe returning a map of parsed arguments.

File: megamek/src/megamek/utilities/ai/BehaviorParameters.java [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+            p7 - other.p7, p8 - other.p8, p9 - other.p9,
+            p10 - other.p10, p11 - other.p11, p12 - other.p12,
+            p13 - other.p13, p14 - other.p14, p15 - other.p15,
+            p16 - other.p16, p17 - other.p17, p18 - other.p18,
+            p19 - other.p19, p20 - other.p20, p21 - other.p21,
+            p22 - other.p22, p23 - other.p23, p24 - other.p24,
+            p25 - other.p25, p26 - other.p26, p27 - other.p27,
+            p28 - other.p28, p29 - other.p29
+        );
+    }
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/BehaviorParameters.java:45]: I fear you're going to be hating life the first time you add another parameter.
Any chance this could be made more flexible, or at least configurable at runtime?

File: megamek/src/megamek/utilities/ai/BehaviorParameters.java [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+        return fromArray(params);
+    }
+
+    public double get(int index) {
+        double[] params = toArray();
+        return params[index];
+    }
+
+    public int size() {
+        return 28;
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/BehaviorParameters.java:119]: Good lord, no :)

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 6]
```diff
@@ -0,0 +1,196 @@
...
+     * Parses a dataset from a file. Can be chained with other parse calls to create a large single training dataset.
+     * @param file The file to parse
+     * @return The parser instance
+     */
+    public DatasetParser parse(File file) {
+        try (BufferedReader reader = new BufferedReader(new FileReader(file))) {
+            String line;
+            while ((line = reader.readLine()) != null) {
+                if (line.startsWith("PLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tFACING\tFROM_X\tFROM_Y\tTO_X\tTO_Y\tHEXES_MOVED\tDISTANCE\tMP_USED\tMAX_MP\tMP_P\tHEAT_P\tARMOR_P\tINTERNAL_P\tJUMPING\tPRONE\tLEGAL\tSTEPS")) {
+                    // Parse action line
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:53]: These need to be an enum or something.

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+                    String actionLine = reader.readLine();
+                    if (actionLine == null) break;
+                    UnitAction action = parseActionLine(actionLine);
+
+                    // Parse state block
+                    line = reader.readLine(); // State header
+                    if (line == null || !line.startsWith("ROUND\tPHASE\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
+                        // || !line.startsWith("ROUND\tPHASE\tTEAM_ID\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:61]: Same here.

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+
+                        if (line == null // || !line.startsWith("ROUND\tPHASE\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
+                            || !line.startsWith("ROUND\tPHASE\tTEAM_ID\tPLAYER_ID\tENTITY_ID\tCHASSIS\tMODEL\tTYPE\tROLE\tX\tY\tFACING\tMP\tHEAT\tPRONE\tAIRBORNE\tOFF_BOARD\tCRIPPLED\tDESTROYED\tARMOR_P\tINTERNAL_P\tDONE")) {
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:64]: Same here.

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+                            throw new RuntimeException("Invalid state header after action: " + line);
+                        }
+
+                    }
+
+                    List<UnitState> states = new ArrayList<>();
+                    Integer currentRound = null;
+                    while ((line = reader.readLine()) != null && (!line.startsWith("PLAYER_ID\tENTITY_ID"))) {
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:72]: Same here.

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+            int mpUsed = Integer.parseInt(parts[11]);
+            int maxMp = Integer.parseInt(parts[12]);
+            double mpP = Double.parseDouble(parts[13]);
+            double heatP = Double.parseDouble(parts[14]);
+            double armorP = Double.parseDouble(parts[15]);
+            double internalP = Double.parseDouble(parts[16]);
+            boolean jumping = parts[17].equals("1");
+            boolean prone = parts[18].equals("1");
+            boolean legal = parts[19].equals("1");
+            // Need to check if the unit in the previous turn had moved or not
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:133]: I highly, highly recommend replacing all of these indexes with constants or enum values or something.

File: megamek/src/megamek/utilities/ai/DatasetParser.java [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+            double mp = Double.parseDouble(parts[index++]);
+            double heat = Double.parseDouble(parts[index++]);
+            boolean prone = parts[index++].equals("1");
+            boolean airborne = parts[index++].equals("1");
+            boolean offBoard = parts[index++].equals("1");
+            boolean crippled = parts[index++].equals("1");
+            boolean destroyed = parts[index++].equals("1");
+            double armorP = Double.parseDouble(parts[index++]);
+            double internalP = Double.parseDouble(parts[index++]);
+            boolean done = parts[index].equals("1");
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/DatasetParser.java:172]: This looks incredibly brittle.  Also, are you modifying the data to indicate that you're done reading the data?  I am not a fan of this.

File: megamek/src/megamek/utilities/ai/ExtendedCostFunction.java [hunk 12]
```diff
@@ -0,0 +1,158 @@
...
+        double curiosityTerm     = computeCuriosityTerm(nextUnitState.get(unitAction.id()));
+
+        // Weight for primary objective (victory)
+        double alpha = behaviorParameters.p23();
+        // Weight for resource-related term
+        double beta = behaviorParameters.p24();
+        // Weight for board-control term
+        double gamma = behaviorParameters.p25();
+        // Weight for exploration/curiosity term
+        double curiosityW = behaviorParameters.p26();
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/ExtendedCostFunction.java:48]: I really think making these parameters named fields or something would be more maintainable going forward.

File: megamek/src/megamek/utilities/ai/ParameterOptimizer.java [hunk 13]
```diff
@@ -0,0 +1,327 @@
...
+
+    private final int maxIterations; // = 100_000;
+    private final double tolerance; // = 1e-6;
+    private final double baseLearningRate; // = 1e-3;
+    private final double maxLearningRate; // = 1e-1;
+    private final int cycleLength; // = 2000;
+
+    private double learningRate;
+    private final int patience; // = 50;
+    private BehaviorParameters velocity = new BehaviorParameters(0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0);
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/ParameterOptimizer.java:39]: You're definitely going to want to make this its own constructor.

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java [hunk 14]
```diff
@@ -0,0 +1,1204 @@
...
+    }
+
+    private static double getWaterCoverScore(BehaviorParameters params, Hex targetHex) {
+        if (targetHex.hasDepth1WaterOrDeeper()) {
+            return 0.3 * params.p15();
+        }
+        return 0.0;
+    }
+
+    private double calculateBuildingCovert(BehaviorParameters params, Coords targetPosition) {
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:264]: Should this be `calculateBuildingCover`?  Or is it "Covert" for being hidden?

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java [hunk 15]
```diff
@@ -0,0 +1,1204 @@
...
+        }
+        return health * weight;
+    }
+
+    private double calculateSelfPreservationMod(UnitAction unitAction, Map<Integer, UnitState> currentUnitStates, BehaviorParameters behaviorParameters) {
+        double weight = behaviorParameters.p6();
+        double health = (unitAction.armorP() + unitAction.internalP()) / 2;
+        if (health < 0.7) {
+            weight *= 1.5; // More cautious when damaged
+        }
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:630]: Is there a reason for this to be discrete rather than, say, using the lerp function to produce a continuous range?

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java [hunk 16]
```diff
@@ -0,0 +1,1204 @@
...
+            Terrains.MAGMA,
+            Terrains.ICE,
+            Terrains.WATER,
+            Terrains.BUILDING,
+            Terrains.BRIDGE,
+            Terrains.BLACK_ICE,
+            Terrains.SNOW,
+            Terrains.SWAMP,
+            Terrains.MUD,
+            Terrains.TUNDRA));
```
Comment (Sleet01) [inline megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:1089]: Could re-use the previously defined set.

**Review by Sleet01 — CHANGES_REQUESTED** [review]
There are a few sections that need work, although overall this looks really promising.

I'd be interested to see the cube coordinates integrated into the main game; as I recently found, there are a number of pain points with our existing offset 2D grid implementation.

File: megamek/src/megamek/common/Entity.java (hunk 2 above)
Comment (Scoppio) [reply megamek/src/megamek/common/Entity.java:2577]: Not rather, but would be nice to have a lightweight game.
Entity today expects that `Game` can be null in some situations, but does not handle `null` in all the rest, we need to make sure that either it is impossible to have Entity without a reference to Game, or make Entity be able to exists witohut `Game`.

File: megamek/src/megamek/utilities/PrincessFineTuning.java (hunk 3 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/PrincessFineTuning.java:58]: I think it is a fair ask. I would prefer not to add a new library in this pull request, but I can do that with no problem.

File: megamek/src/megamek/utilities/ai/BehaviorParameters.java (hunk 4 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/BehaviorParameters.java:45]: Copilot kinda automatize this for me, but you are right. I will update this to just use an `double[]` instead.

File: megamek/src/megamek/utilities/ai/BehaviorParameters.java (hunk 5 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/BehaviorParameters.java:119]: That explains alot auehhuaehueahueahueauh

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 6 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:53]: I see no benefit into making this an enum, care to elaborate?

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 8 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:64]: I see no benefit into making this an enum, care to elaborate?

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 10 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:133]: Ok, ok... I will make a proper parser.

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java (hunk 14 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:264]: Should be cover

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java (hunk 15 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:630]: No, there is none, I reached a point where I started to throw things at the wall to see what sticks in the hope that I can make the loss of the cost function reach low enough.

File: megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java (hunk 16 above)
Comment (Scoppio) [reply megamek/src/megamek/utilities/ai/UtilityPathRankerCostFunction.java:1089]: We could, I will move it to inside the hex then.

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 6 above)
Comment (Sleet01) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:53]: Oh, I see the issue: I was thinking this was checking for matches from a set, not for a single very long line with tab separation.
At any rate, this should definitely be a constant (something like "ACTION_LINE_HEADER") for clarity and ease of editing later.

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 7 above)
Comment (Sleet01) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:61]: Yeah, apologies, I was thinking this was a set of matching strings, not one huge string.
It would still benefit from being made into a constant for clarity.

File: megamek/src/megamek/utilities/ai/DatasetParser.java (hunk 8 above)
Comment (Sleet01) [reply megamek/src/megamek/utilities/ai/DatasetParser.java:64]: Yeah, apologies, I was thinking this was a set of matching strings, not one huge string.
It would still benefit from being made into a constant for clarity.

File: megamek/src/megamek/ai/dataset/TrainingDataset.java [hunk 17]
```diff
@@ -1,10 +1,19 @@
-package megamek.utilities.ai;
+package megamek.ai.dataset;
```
Comment (Sleet01) [inline megamek/src/megamek/ai/dataset/TrainingDataset.java:1]: Copyright header ^^^

File: megamek/src/megamek/ai/dataset/TsvSerde.java [hunk 18]
```diff
@@ -0,0 +1,7 @@
+package megamek.ai.dataset;
```
Comment (Sleet01) [inline megamek/src/megamek/ai/dataset/TsvSerde.java:1]: ^^^ Copyright header

File: megamek/src/megamek/ai/optimizer/CostFunctionChooser.java [hunk 19]
```diff
@@ -0,0 +1,33 @@
+package megamek.ai.optimizer;
```
Comment (Sleet01) [inline megamek/src/megamek/ai/optimizer/CostFunctionChooser.java:14]: Copyright header ^^^

Comment (Scoppio) [reply megamek/src/megamek/ai/optimizer/CostFunctionChooser.java:14]: already fixed

**Review by Sleet01 — APPROVED** [review]
A few files are missing Copyright headers but that doesn't impact functionality.
This looks great!
